File: client/python/krpc/test/test_service_definitions.py
import unittest
import warnings
from inspect import signature
from typing import Dict, List, NamedTuple
from krpc.client import _stub_definitions
from krpc.definitions import register_all
from krpc.encoder import Encoder
from krpc.service import create_service, service_definitions
from krpc.types import Types
import krpc.schema.KRPC_pb2 as KRPC


class FakeClient:
    """As much of a client as creating a service needs, without a connection to a server"""

    def __init__(self) -> None:
        self._types = Types()

    def _invoke(self, *args: object, **kwargs: object) -> object:
        raise NotImplementedError

    def _build_call(self, *args: object, **kwargs: object) -> object:
        raise NotImplementedError


def enum_service(name: str) -> KRPC.Service:
    """A service defining an enumeration and nothing else"""
    service = KRPC.Service()
    service.name = name
    enumeration = service.enumerations.add()
    enumeration.name = "Mode"
    for value_name, value in (("Slow", 1), ("Fast", 2)):
        enumeration_value = enumeration.values.add()
        enumeration_value.name = value_name
        enumeration_value.value = value
    return service


def default_service(
    name: str, enum_service_name: str, in_a_list: bool = False
) -> KRPC.Service:
    """A service with a procedure whose parameter defaults to a member of another service's
    enumeration, either directly or through a list"""
    types = Types()
    enumeration_type = types.enumeration_type(enum_service_name, "Mode")
    encoded = Encoder.encode(2, types.sint32_type)
    if in_a_list:
        items = KRPC.List()
        items.items.extend([encoded])
        encoded = items.SerializeToString()

    service = KRPC.Service()
    service.name = name
    procedure = service.procedures.add()
    procedure.name = "Frob"
    parameter = procedure.parameters.add()
    parameter.name = "mode"
    if in_a_list:
        parameter.type.CopyFrom(types.list_type(enumeration_type).protobuf_type)
    else:
        parameter.type.CopyFrom(enumeration_type.protobuf_type)
    parameter.has_default_value = True
    parameter.default_value = encoded
    return service


# A type code from a later version of the protocol than this client knows about
UNKNOWN_TYPE_CODE = 9999


def struct_service(name: str, field_type: KRPC.Type) -> KRPC.Service:
    """A service defining a structure with a single field of the given type"""
    service = KRPC.Service()
    service.name = name
    struct = service.structs.add()
    struct.name = "Thing"
    field = struct.fields.add()
    field.name = "Value"
    field.type.CopyFrom(field_type)
    return service


def struct_using_service(
    name: str, struct_service_name: str, with_default: bool = False
) -> KRPC.Service:
    """A service with a procedure taking a parameter of another service's structure,
    optionally one with a default value"""
    types = Types()
    struct_type = types.struct_type(struct_service_name, "Thing")

    service = KRPC.Service()
    service.name = name
    procedure = service.procedures.add()
    procedure.name = "Frob"
    parameter = procedure.parameters.add()
    parameter.name = "thing"
    parameter.type.CopyFrom(struct_type.protobuf_type)
    if with_default:
        parameter.has_default_value = True
        parameter.default_value = KRPC.Tuple().SerializeToString()
    return service


def unknown_type() -> KRPC.Type:
    typ = KRPC.Type()
    typ.code = UNKNOWN_TYPE_CODE
    return typ


class TestServiceDefinitions(unittest.TestCase):
    @staticmethod
    def create_services(services: List[KRPC.Service]) -> FakeClient:
        client = FakeClient()
        definitions = [
            definition
            for service in services
            for definition in service_definitions(service)
        ]
        register_all(client._types, definitions)
        for service in services:
            setattr(client, service.name, create_service(client, service))  # type: ignore[arg-type]
        return client

    def check_default(self, client: FakeClient, expected: str) -> None:
        service = getattr(client, "ServiceA")
        default = signature(service.frob).parameters["mode"].default
        self.assertEqual(expected, str(default))

    def test_defined_before_it_is_used(self) -> None:
        client = self.create_services(
            [enum_service("ServiceB"), default_service("ServiceA", "ServiceB")]
        )
        self.check_default(client, "Mode.fast")

    def test_defined_after_it_is_used(self) -> None:
        client = self.create_services(
            [default_service("ServiceA", "ServiceB"), enum_service("ServiceB")]
        )
        self.check_default(client, "Mode.fast")

    def test_used_inside_a_collection(self) -> None:
        client = self.create_services(
            [
                default_service("ServiceA", "ServiceB", in_a_list=True),
                enum_service("ServiceB"),
            ]
        )
        self.check_default(client, "[<Mode.fast: 2>]")

    def test_defined_by_the_service_that_uses_it(self) -> None:
        service = default_service("ServiceA", "ServiceA")
        service.enumerations.extend(enum_service("ServiceA").enumerations)
        self.check_default(self.create_services([service]), "Mode.fast")

    def check_struct_is_skipped(self, services: List[KRPC.Service]) -> None:
        with warnings.catch_warnings(record=True) as caught:
            warnings.simplefilter("always")
            client = self.create_services(services)
        messages = [str(warning.message) for warning in caught]
        self.assertTrue(any("Thing" in message for message in messages), messages)
        self.assertTrue(
            any("ServiceA.Frob" in message for message in messages), messages
        )
        self.assertFalse(hasattr(getattr(client, "ServiceA"), "frob"))

    def test_procedure_naming_a_skipped_struct_is_skipped(self) -> None:
        self.check_struct_is_skipped(
            [
                struct_service("ServiceB", unknown_type()),
                struct_using_service("ServiceA", "ServiceB"),
            ]
        )

    def test_defaulted_parameter_of_a_skipped_struct_is_skipped(self) -> None:
        self.check_struct_is_skipped(
            [
                struct_service("ServiceB", unknown_type()),
                struct_using_service("ServiceA", "ServiceB", with_default=True),
            ]
        )

    def test_struct_holding_a_skipped_struct_is_skipped(self) -> None:
        types = Types()
        skipped = struct_service("ServiceC", unknown_type())
        holder = struct_service(
            "ServiceB", types.struct_type("ServiceC", "Thing").protobuf_type
        )
        self.check_struct_is_skipped(
            [skipped, holder, struct_using_service("ServiceA", "ServiceB")]
        )

    def test_stub_with_more_fields_than_the_server(self) -> None:
        # An older server declares fewer fields than the stubs were generated against, so a
        # value of the structure cannot be built as the type the stubs provide
        class Thing(NamedTuple):
            value: int
            appended: str

        class StubService:
            _classes: Dict[str, type] = {}
            _enumerations: Dict[str, type] = {}
            _structs = {"Thing": Thing}

        client = FakeClient()
        service = struct_service("ServiceB", Types().sint32_type.protobuf_type)
        with warnings.catch_warnings(record=True) as caught:
            warnings.simplefilter("always")
            register_all(
                client._types, _stub_definitions(client, service, StubService())
            )
        self.assertTrue(
            any("ServiceB.Thing" in str(warning.message) for warning in caught)
        )
        struct_type = client._types.struct_type("ServiceB", "Thing")
        self.assertEqual(["value"], struct_type.field_names)
        self.assertIsNot(Thing, struct_type.python_type)

    def test_defined_by_no_service(self) -> None:
        with self.assertRaises(RuntimeError) as cm:
            self.create_services([default_service("ServiceA", "ServiceB")])
        self.assertIn("ServiceA.Frob parameter mode", str(cm.exception))
        self.assertIn("ServiceB.Mode", str(cm.exception))


if __name__ == "__main__":
    unittest.main()

File: client/lua/krpc/test/test_service_definitions.lua
local luaunit = require 'luaunit'
local class = require 'pl.class'
local List = require 'pl.List'
local schema = require 'krpc.schema.KRPC'
local Types = require 'krpc.types'
local encoder = require 'krpc.encoder'
local service = require 'krpc.service'

local TestServiceDefinitions = class()

--- As much of a client as creating a service needs, without a connection to a server
local FakeClient = class()

function FakeClient:_init()
  self._types = Types()
end

function FakeClient:_invoke()
  error('not connected to a server')
end

--- A service defining an enumeration and nothing else
local function enum_service(name)
  local svc = schema.Service()
  svc.name = name
  local enum = svc.enumerations:add()
  enum.name = 'Mode'
  for _,case in ipairs({{'Slow', 1}, {'Fast', 2}}) do
    local value = enum.values:add()
    value.name = case[1]
    value.value = case[2]
  end
  return svc
end

--- A service with a procedure whose parameter defaults to a member of another service's
--- enumeration, either directly or through a list
local function default_service(name, enum_service_name, in_a_list)
  local types = Types()
  local svc = schema.Service()
  svc.name = name
  local procedure = svc.procedures:add()
  procedure.name = 'Frob'
  local parameter = procedure.parameters:add()
  parameter.name = 'mode'
  local encoded = encoder.encode(2, types:sint32_type())
  if in_a_list then
    parameter.type.code = Types.LIST
    local value_type = parameter.type.types:add()
    value_type.code = Types.ENUMERATION
    value_type.service = enum_service_name
    value_type.name = 'Mode'
    local items = schema.List()
    items.items:append(encoded)
    encoded = items:SerializeToString()
  else
    parameter.type.code = Types.ENUMERATION
    parameter.type.service = enum_service_name
    parameter.type.name = 'Mode'
  end
  parameter.has_default_value = true
  parameter.default_value = encoded
  return svc
end

-- A type code from a later version of the protocol than this client knows about
local UNKNOWN_TYPE_CODE = 9999

--- A service defining a structure with a single field, whose type the given function sets
local function struct_service(name, set_field_type)
  local svc = schema.Service()
  svc.name = name
  local struct = svc.structs:add()
  struct.name = 'Thing'
  local field = struct.fields:add()
  field.name = 'Value'
  set_field_type(field.type)
  return svc
end

--- A service with a procedure taking a parameter of another service's structure
local function struct_using_service(name, struct_service_name)
  local svc = schema.Service()
  svc.name = name
  local procedure = svc.procedures:add()
  procedure.name = 'Frob'
  local parameter = procedure.parameters:add()
  parameter.name = 'thing'
  parameter.type.code = Types.STRUCT
  parameter.type.service = struct_service_name
  parameter.type.name = 'Thing'
  return svc
end

local function unknown_field_type(typ)
  typ.code = UNKNOWN_TYPE_CODE
end

local function struct_field_type(service_name)
  return function (typ)
    typ.code = Types.STRUCT
    typ.service = service_name
    typ.name = 'Thing'
  end
end

local function create_services(svcs)
  local client = FakeClient()
  for _,svc in ipairs(svcs) do
    service.register_definitions(client._types, svc)
  end
  local created = {}
  for _,svc in ipairs(svcs) do
    created[svc.name] = service.create_service(client, svc)
  end
  return client, created
end

--- Create the given services, and return the client along with the default value that
--- ServiceA decoded for the parameter of its procedure
local function create_services_and_default(svcs)
  local client, created = create_services(svcs)
  luaunit.assertNotNil(created['ServiceA'].frob)
  local procedure
  for _,svc in ipairs(svcs) do
    if svc.name == 'ServiceA' then
      procedure = svc.procedures[1]
    end
  end
  local _,_,_,param_default = created['ServiceA']:_parse_procedure(procedure)
  return client, param_default[1]
end

function TestServiceDefinitions:check_services(svcs, in_a_list)
  local client, default = create_services_and_default(svcs)
  local fast = client._types:enumeration_type('ServiceB', 'Mode').lua_type.fast
  luaunit.assertEquals(2, fast.value)
  if in_a_list then
    luaunit.assertEquals(List{fast}, default)
  else
    luaunit.assertEquals(fast, default)
  end
end

function TestServiceDefinitions:test_defined_before_it_is_used()
  self:check_services({enum_service('ServiceB'), default_service('ServiceA', 'ServiceB')})
end

function TestServiceDefinitions:test_defined_after_it_is_used()
  self:check_services({default_service('ServiceA', 'ServiceB'), enum_service('ServiceB')})
end

function TestServiceDefinitions:test_used_inside_a_collection()
  self:check_services({default_service('ServiceA', 'ServiceB', true), enum_service('ServiceB')}, true)
end

function TestServiceDefinitions:check_struct_is_skipped(svcs)
  local _, created = create_services(svcs)
  luaunit.assertNil(created['ServiceA'].frob)
end

function TestServiceDefinitions:test_procedure_naming_a_skipped_struct_is_skipped()
  local _, created = create_services({
    struct_service('ServiceB', unknown_field_type),
    struct_using_service('ServiceA', 'ServiceB')
  })
  luaunit.assertNil(created['ServiceA'].frob)
  luaunit.assertNil(created['ServiceB'].Thing)
end

function TestServiceDefinitions:test_struct_holding_a_skipped_struct_is_skipped()
  self:check_struct_is_skipped({
    struct_service('ServiceC', unknown_field_type),
    struct_service('ServiceB', struct_field_type('ServiceC')),
    struct_using_service('ServiceA', 'ServiceB')
  })
end

return TestServiceDefinitions
